In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from sklearn.preprocessing import StandardScaler
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

In [2]:
df_train = pd.read_csv("train_KNN.csv")
df_test = pd.read_csv("test_KNN.csv")

df_train.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


In [3]:
df_train["Age"] = df_train["Age"].fillna(df_train["Age"].median())
df_test["Age"] = df_test["Age"].fillna(df_test["Age"].median())

df_train = df_train.dropna(subset=["Embarked"])
df_test = df_test.dropna(subset=["Fare"])

df_train = df_train.drop(["Cabin", "Name", "Ticket", "PassengerId"], axis=1)
df_test = df_test.drop(["Cabin", "Name", "Ticket", "PassengerId"], axis=1)

In [4]:
df_train = pd.get_dummies(df_train, drop_first=True)
df_test = pd.get_dummies(df_test, drop_first=True)

In [5]:
X_train = df_train.drop("Survived", axis=1)
y_train = df_train["Survived"]

X_test = df_test.drop("Survived", axis=1)
y_test = df_test["Survived"]

In [6]:
scaler = StandardScaler()

X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

In [7]:
knn = KNeighborsClassifier(n_neighbors=5)
knn.fit(X_train, y_train)

y_pred = knn.predict(X_test)

knn_accuracy = accuracy_score(y_test, y_pred)
print("KNN Accuracy:", knn_accuracy)

KNN Accuracy: 0.8129496402877698


In [8]:
lr = LogisticRegression(max_iter=1000)
lr.fit(X_train, y_train)

y_pred_lr = lr.predict(X_test)

lr_accuracy = accuracy_score(y_test, y_pred_lr)
print("Logistic Regression Accuracy:", lr_accuracy)

Logistic Regression Accuracy: 0.9424460431654677


In [9]:
dt = DecisionTreeClassifier()
dt.fit(X_train, y_train)

y_pred_dt = dt.predict(X_test)

dt_accuracy = accuracy_score(y_test, y_pred_dt)
print("Decision Tree Accuracy:", dt_accuracy)

Decision Tree Accuracy: 0.7961630695443646


In [10]:
print("KNN Accuracy:", knn_accuracy)
print("Logistic Regression Accuracy:", lr_accuracy)
print("Decision Tree Accuracy:", dt_accuracy)

KNN Accuracy: 0.8129496402877698
Logistic Regression Accuracy: 0.9424460431654677
Decision Tree Accuracy: 0.7961630695443646


## Findings and Interpretation

In this experiment, three machine learning algorithms were applied to the Titanic dataset to predict passenger survival.

The K-Nearest Neighbors (KNN) model achieved an accuracy of approximately 81%, showing moderate performance. Logistic Regression performed the best with an accuracy of around 94%, indicating that it is the most suitable model for this dataset. The Decision Tree model achieved around 79% accuracy, which is slightly lower and may be affected by overfitting.

From the comparison, it is clear that Logistic Regression provides the most reliable results, while KNN and Decision Tree show comparatively lower performance.

This experiment demonstrates that different machine learning models produce different results, and choosing the right model is important for achieving high accuracy.
